src/id128.rs

```diff
@@ -53,83 +44,130 @@ impl Id128 {
+        let buf = fs::read_to_string("/etc/machine-id")
+            .map_err(|e| format!("failed to open machine-id: {}", e))?;
+        Id128::from_str(buf.trim_end())
     }
 
     /// Return this ID as a lowercase hexadecimal string, without dashes.
     pub fn lower_hex(&self) -> String {
-        let mut hex = String::new();
-        for byte in self.uuid_v4.as_bytes() {
-            hex.push_str(&format!("{:02x}", byte));
-        }
-        hex
+        self.0.to_simple_ref().to_string()
     }
 
     /// Return this ID as a lowercase hexadecimal string, with dashes.
     pub fn dashed_hex(&self) -> String {
-        format!("{}", self.uuid_v4.to_hyphenated_ref())
+        self.0.to_hyphenated_ref().to_string()
     }
+}
+
+impl AsRef<[u8]> for Id128 {
+    fn as_ref(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
+
+impl AsRef<OsStr> for Id128 {
+    fn as_ref(&self) -> &OsStr {
+        OsStr::from_bytes(self.as_bytes())
+    }
+}
+
+impl From<[u8; 16]> for Id128 {
+    fn from(bytes: [u8; 16]) -> Self {
+        Self(Uuid::from_bytes(bytes))
+    }
+}
+
+impl FromStr for Id128 {
+    type Err = SdError;
 
-    /// Custom serialization (lower hex).
-    fn ser_uuid<S>(field: &Uuid, s: S) -> ::std::result::Result<S::Ok, S::Error>
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        Ok(Self(input.parse().map_err(|e| {
+            format!("failed to parse ID from string: {}", e)
+        })?))
+    }
+}
+
+impl TryFrom<&[u8]> for Id128 {
+    type Error = SdError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        Ok(Self(Uuid::from_slice(bytes).map_err(|e| {
+            format!("failed to parse ID from bytes slice: {}", e)
+        })?))
+    }
+}
+
+// TODO: (in 0.4): make optional behind serde feature
+impl serde_crate::Serialize for Id128 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
-        S: ::serde::Serializer,
+        S: serde_crate::Serializer,
     {
-        let mut hex = String::new();
-        for byte in field.as_bytes() {
-            hex.push_str(&format!("{:02x}", byte));
+        if serializer.is_human_readable() {
+            self.lower_hex().serialize(serializer)
+        } else {
+            self.as_bytes().serialize(serializer)
         }
-        s.serialize_str(&hex)
     }
 }
 
-impl fmt::Debug for Id128 {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self.dashed_hex())
+impl<'de> serde_crate::Deserialize<'de> for Id128 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde_crate::Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            Ok(Self(Uuid::deserialize(deserializer)?))
+        } else {
+            Ok(Self(uuid::adapter::compact::deserialize(deserializer)?))
+        }
     }
 }
 
-/// Return this machine unique ID.
+#[deprecated(since = "0.3.2", note = "use Id128::from_machine()")]
```

Comment:
>
> 
> The current naming may look ugly, but it is meant to stick closer to the C `sd_id128_*` APIs.

Yeah that was my understanding reading the code too. It's just that the C API is... for C so since this is a Rust lib it should follow rust API guidelines. Keeping C API is confusing

Reply:
True... to a certain extent. Keeping some alignment on the names helps a lot with discoverability, that's what I've see from other people/consumers using it.
My rule of thumb is roughly to keep (very similar / identical) names for things that are basically 1:1 ports, and exercise a bit more freedom for things that do not fall in that bucket. Here the former applies though.

Reply:
Then we can keep the name (`get_machine()`) but still stick it onto the `Id128` struct. By not having free functions we can in 0.4 hide the id128 module and just reexport the `Id128` struct.

0.3: `libsystemd::id128::get_machine()`
0.4 `libsystemd::Id128::get_machine()`

Looking at https://github.com/coreos/zincati/blob/10a9c409013364697146ff5bf43c8bdaede347ae/src/identity/mod.rs having the `id128` module makes the code much uglier than if all the functions were on the `Id128` struct.